Allow collecting history of query execution - requests, retries, speculative fibers

cvybhu opened the pull request:
This PR makes it possible to collect execution history of a query.
This history contains information about the way a query was executed by session.
This includes requests to specific nodes, retry decisions, whether the attempt was a part of speculative execution, etc.

This information can be obtained by implementing the `HistoryListener` trait and passing it to `StatementConfig`:
```rust
pub trait HistoryListener: Debug + Send + Sync {
    /// Log that a query has st on query start - right after the call to Session::query.
    fn log_query_start(&self) -> QueryId;

    /// Log that query was successfull - called right before returning the result from Session::query.
    fn log_query_success(&self, query_id: QueryId);

    /// Log that query ended with an error - called right before returning the error from Session::query.
    fn log_query_error(&self, query_id: QueryId, error: &QueryError);

    /// Log that a new speculative fiber has started.
    fn log_new_speculative_fiber(&self, query_id: QueryId) -> SpeculativeId;

    /// Log that an attempt has started - request has been sent on some Connection, now awaiting for an answer.
    fn log_attempt_start(
        &self,
        query_id: QueryId,
        speculative_id: Option<SpeculativeId>,
        node_addr: SocketAddr,
    ) -> AttemptId;

    /// Log that an attempt succeded.
    fn log_attempt_success(&self, attempt_id: AttemptId);

    /// Log that an attempt ended with an error. The error and decision whether to retry the attempt are also included in the log.
    fn log_attempt_error(
        &self,
        attempt_id: AttemptId,
        error: &QueryError,
        retry_decision: &RetryDecision,
    );
}
```

### TODOs
The code is ready for review, but there are still some related things that haven't been done yet:
- [ ]  Add a chapter in the documentation book
- [ ] Add basic tests - more complex tests would have to wait for Proxy, but some simple ones should be added.
- [ ] Run some manual tests - it seems to work for the basic cases, but it would be good to manually set up a Proxy and test some more complex ones.

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [ ] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

cvybhu commented:
Here's an example output for an iterator which fetched 4 pages:
```
Iterator history: Queries History:
=== Query #0 ===
| start_time: 2022-08-25 11:21:50.445075147 UTC
| Non-speculative attempts:
| - Attempt #0 sent to 127.0.0.1:9042
|   request send time: 2022-08-25 11:21:50.445151628 UTC
|   Success at 2022-08-25 11:21:50.447444362 UTC
|
| Query succesful at 2022-08-25 11:21:50.447447970 UTC
=================
=== Query #1 ===
| start_time: 2022-08-25 11:21:50.447464514 UTC
| Non-speculative attempts:
| - Attempt #0 sent to 127.0.0.1:9042
|   request send time: 2022-08-25 11:21:50.447476923 UTC
|   Success at 2022-08-25 11:21:50.449853729 UTC
|
| Query succesful at 2022-08-25 11:21:50.449858973 UTC
=================
=== Query #2 ===
| start_time: 2022-08-25 11:21:50.449882873 UTC
| Non-speculative attempts:
| - Attempt #0 sent to 127.0.0.1:9042
|   request send time: 2022-08-25 11:21:50.449889715 UTC
|   Success at 2022-08-25 11:21:50.452576455 UTC
|
| Query succesful at 2022-08-25 11:21:50.452580846 UTC
=================
=== Query #3 ===
| start_time: 2022-08-25 11:21:50.452600988 UTC
| Non-speculative attempts:
| - Attempt #0 sent to 127.0.0.1:9042
|   request send time: 2022-08-25 11:21:50.452604371 UTC
|   Success at 2022-08-25 11:21:50.455373461 UTC
|
| Query succesful at 2022-08-25 11:21:50.455376847 UTC
=================
=== Query #4 ===
| start_time: 2022-08-25 11:21:50.455394558 UTC
| Non-speculative attempts:
| - Attempt #0 sent to 127.0.0.1:9042
|   request send time: 2022-08-25 11:21:50.455405591 UTC
|   Success at 2022-08-25 11:21:50.456763674 UTC
|
| Query succesful at 2022-08-25 11:21:50.456765671 UTC
=================
```

psarna commented:
I only did a cursory review for now - looks nice in general, the code is sometimes a little boilerplate-y, but I guess it needs to be that way since the events are similar to each other. The output looks neat too

cvybhu commented:
v2:
* s/history_stuff/history_listener_and_id
* s/his_data/history_data
* s/cur/current
* Improved newlines in querey history Display impl
* Added a chapter in the documentation book
* Added unit tests for query history

cvybhu commented:
I will try to setup Proxy and try to run a few manual tests, trigger speculative execution, etc. and see if this is collected correctly.

cvybhu commented:
Alright I painfully rebased this PR on top of Proxy and added history logging to the test that triggers speculative execution ([here's the branch](https://github.com/cvybhu/scylla-rust-driver/tree/proxy_with_history)).
Looks like everything works alright. I can see the separate speculative fibers and their attempts:
```none
=== Query #0 ===
| start_time: 2022-08-26 15:08:28.525367409 UTC
| Non-speculative attempts:
| - Attempt #0 sent to 127.0.0.219:9042
|   request send time: 2022-08-26 15:08:28.525409294 UTC
|   No result yet
|
|
| > Speculative fiber #0
| fiber start time: 2022-08-26 15:08:28.537074167 UTC
| - Attempt #0 sent to 127.0.0.217:9042
|   request send time: 2022-08-26 15:08:28.537126083 UTC
|   No result yet
|
|
| > Speculative fiber #1
| fiber start time: 2022-08-26 15:08:28.548050242 UTC
| - Attempt #0 sent to 127.0.0.218:9042
|   request send time: 2022-08-26 15:08:28.548089083 UTC
|   Success at 2022-08-26 15:08:28.590052778 UTC
|
| Query succesful at 2022-08-26 15:08:28.590078119 UTC
=================
```

One interesting thing is that our speculative execution is very eager. Even with all nodes in perfect condition two more speculative requests were sent before receiving response for the first one x.x:

```none
2022-08-26T15:16:15.751254Z  INFO retries: --------------------- BEGINNING main test part ----------------
2022-08-26T15:16:15.751284Z  INFO retries: --------------------- first query - no rules  ----------------
2022-08-26T15:16:15.834354Z  INFO retries: History: Queries History:
=== Query #0 ===
| start_time: 2022-08-26 15:16:15.751315676 UTC
| Non-speculative attempts:
| - Attempt #0 sent to 127.0.0.219:9042
|   request send time: 2022-08-26 15:16:15.751370307 UTC
|   Success at 2022-08-26 15:16:15.834212330 UTC
|
|
| > Speculative fiber #0
| fiber start time: 2022-08-26 15:16:15.762701089 UTC
| - Attempt #0 sent to 127.0.0.217:9042
|   request send time: 2022-08-26 15:16:15.762753439 UTC
|   No result yet
|
|
| > Speculative fiber #1
| fiber start time: 2022-08-26 15:16:15.773206156 UTC
| - Attempt #0 sent to 127.0.0.218:9042
|   request send time: 2022-08-26 15:16:15.773265805 UTC
|   No result yet
|
| Query succesful at 2022-08-26 15:16:15.834236528 UTC
=================
```
Maybe we could consider increasing the default speculative execution timeout so that we don't spam the database this much.

cvybhu commented:
Manual test passed, I feel it's now ready.

cvybhu commented:
> Maybe we could consider increasing the default speculative execution timeout so that we don't spam the database this much.

Or maybe it's just because of the latency that Proxy causes. The request here took ~80ms, that's really long.

wprzytula commented on scylla/src/history.rs:
```diff
@@ -203,3 +219,169 @@ impl HistoryListener for HistoryCollector {
+                    match attempts.get_mut(attempt_id) {
+                        Some(attempt) => {
+                            if attempt.result.is_some() {
+                                warn!("StructuredHistory - atempt with id {:?} has many results", attempt_id);
```
attempt, has multiple results

wprzytula commented on scylla/src/history.rs:
```diff
@@ -203,3 +219,169 @@ impl HistoryListener for HistoryCollector {
+        // Move attempts to their speculative fibers
+        for (event, _) in &data.events {
+            if let HistoryEvent::NewAttempt(attempt_id, query_id, speculative_id, _) = event {
+                let attempt = match attempts.remove(attempt_id) {
+                    Some(a) => a,
+                    None => continue,
+                };
+
+                match speculative_id {
+                    Some(spec_id) => {
+                        if let Some(spec_fiber) = fibers.get_mut(spec_id) {
+                            spec_fiber.attempts.push(attempt);
+                        }
+                    }
+                    None => {
+                        if let Some(query) = queries.get_mut(query_id) {
+                            query.non_speculative_fiber.attempts.push(attempt);
+                        }
+                    }
+                }
```
Less imperative style here would be cleaner IMO than using `continue`:
```
if let Some(attempt) = attempts.remove(attempt_id) {
...
}
```

cvybhu replied:
I wanted to avoid having too many levels of indentation.
As a great man once said:

> [...] if you need more than 3 levels of indentation, you’re screwed anyway, and should fix your program.

I feel that with if let this code would become much harder to parse.

wprzytula replied:
4 levels of indentation are no longer a problem with such big screens we have, whereas overlooking a `continue` is much more likely and leading to confusion. I would strongly advice to avoid using `return` and `continue` in places like that.

wprzytula replied:
That's the reason similar to why GOTOs were deprecated - the code flow with jumps is harder to follow than a linear one.

cvybhu replied:
I feel like the pattern
```
for () {
    if (not condition) {
       continue;
    }

   do stuff
}
```

Is common enough that it shouldn't cause much confusion. `if let` causes there to be a chain of indented ifs :/

But I don't feel that strongly about it, changed it to avoid bikeshedding.

wprzytula commented on scylla/src/history.rs:
```diff
@@ -203,3 +219,169 @@ impl HistoryListener for HistoryCollector {
+        // Move speculative fibers to their queries
+        for (event, _) in &data.events {
+            if let HistoryEvent::NewSpeculativeFiber(speculative_id, query_id) = event {
+                let fiber: FiberHistory = match fibers.remove(speculative_id) {
+                    Some(f) => f,
+                    None => continue,
+                };
+
+                if let Some(query) = queries.get_mut(query_id) {
+                    query.speculative_fibers.push(fiber);
+                }
```
ditto

wprzytula commented on scylla/src/history.rs:
```diff
@@ -382,6 +382,53 @@ impl From<&HistoryCollectorData> for StructuredHistory {
+            writeln!(f, "|")?;
+            match &query.result {
+                Some(QueryHistoryResult::Success(succ_time)) => {
+                    writeln!(f, "| Query succesful at {}", succ_time)?;
```
successful

wprzytula commented on examples/query_history.rs:
```diff
@@ -0,0 +1,69 @@
+    query.set_history_listener(history_listener.clone());
+
+    // Run the query, doesn't matter if it failed, the history will still be saved
+    let _ignore_eror = session.query(query.clone(), ()).await;
```
ignore_error

wprzytula commented on scylla/src/history.rs:
```diff
@@ -115,6 +121,16 @@ impl HistoryCollector {
 
+    /// Clone the collected events and convert them to StructuredHistory.
+    pub fn clone_structured_history(&self) -> StructuredHistory {
+        StructuredHistory::from(&self.clone_colected())
```
collected

wprzytula commented on examples/query_history.rs:
```diff
@@ -0,0 +1,69 @@
+
+    println!(
+        "Iterator history: {}",
+        iter_history_listener.clone_structured_history()
```
You only use `clone_structured_history()` in the example, while `take_structured_history()` seems to be more reasonable for a usual use (at least is appears so to me - what's the use cloning the history if printing it just once would suffice anyway?). Therefore, I suggest changing this last `clone...` into `take...`.

cvybhu replied:
;+1: Changed `clone` to `take` for the last call,

wprzytula commented on docs/source/tracing/tracing.md:
```diff
@@ -1,8 +1,12 @@
+### Tracing
+
+Tracing is a feature provided by Scylla. When sending a query we can set a flag that signifies that we would like it to be traced.
+After completing the query Scylla provides a `tracing_id` which can be used to fetch information abot it - which nodes it was sent to, what operations were performed etc.
```
about it

wprzytula commented on scylla/src/history.rs:
```diff
@@ -432,3 +432,649 @@ fn write_fiber_attempts(fiber: &FiberHistory, f: &mut std::fmt::Formatter<'_>) -
+    }
+
+    // Set a single node for all attempts within StructuredHistory.
+    // When running agains real life nodes this address may change,
```
against

wprzytula commented on scylla/src/history.rs:
```diff
@@ -432,3 +432,649 @@ fn write_fiber_attempts(fiber: &FiberHistory, f: &mut std::fmt::Formatter<'_>) -
+
+        let _query_id: QueryId = history_collector.log_query_start();
+
+        let history: StructuredHistory = history_collector.clone_structured_history();
```
Why do you prefer to clone instead of taking the structured history?

cvybhu replied:
I didn't want to think whether taking is ok or not, so I just cloned stuff. Those are tests anyway.

wprzytula commented on scylla/src/history.rs:
```diff
@@ -432,3 +432,649 @@ fn write_fiber_attempts(fiber: &FiberHistory, f: &mut std::fmt::Formatter<'_>) -
+| - Attempt #0 sent to 127.0.0.1:19042
+|   request send time: 2022-02-22 20:22:22 UTC
+|   Error at 2022-02-22 20:22:22 UTC
+|   Error: Database returned an error: The submitted query has a syntax error, Error message: Error message from database
```
As @piodul once mentioned, relying on error messages in tests can be pain when they change, but it seems inevitable here.

cvybhu replied:
I try to mitigate this by setting the error message to a constant string using `set_one_db_error_message`. The test will break only once our internal error printing breaks, which will be a code change anyway.

cvybhu commented on scylla/src/history.rs:
```diff
@@ -0,0 +1,60 @@
+
+/// Id of a single query, i.e a single call to Session::query/execute/etc.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Default, Debug)]
+pub struct QueryId(pub usize);
```
It was convenient to have it when implementing Default for CollectedHistoryData, but yeah a bit strange that Id has a default value. removed.

cvybhu commented on scylla/src/transport/iterator.rs:
```diff
@@ -446,6 +466,86 @@ where
+        self.current_query_id = Some(history_listener.log_query_start());
+    }
+
+    fn log_query_success(&mut self) {
+        let history_listener: &dyn HistoryListener = match &self.history_listener {
+            Some(hl) => &**hl,
+            None => return,
+        };
+
+        let query_id: history::QueryId = match &self.current_query_id {
+            Some(id) => *id,
+            None => return,
+        };
+
+        history_listener.log_query_success(query_id);
+    }
```
I like the current version more.
There are three simple steps in the function - take this, take that, do this.
The alternative is one complicated line that requires much more cognitive load to parse.
Unless there are more votes for the alternative I'm gonna stay with the current one.

wprzytula replied:
Well, I guess the amount of cognitive load needed to parse both versions depends on one's preference and experience with more imperative or more functional code style - so it's probably the matter of taste.

cvybhu commented on scylla/src/history.rs:
```diff
@@ -58,3 +61,145 @@ pub trait HistoryListener: Debug + Send + Sync {
+/// Each event is marked with an UTC timestamp.
+#[derive(Debug, Default)]
+pub struct HistoryCollector {
+    data: Mutex<HistoryCollectorData>,
```
I assumed that we don't care about efficiency at all.
This will most likely be used in debugging contexts.

The interface defined by `HistoryListener` trait is generic enough that we can implement a more efficient collector if the need arises.

cvybhu commented on scylla/src/history.rs:
```diff
@@ -58,3 +61,145 @@ pub trait HistoryListener: Debug + Send + Sync {
+    /// It's possible that after finishing a query and taking out the events
+    /// new ones will still come - from queries that haven't been cancelled yet.
+    pub fn take_collected(&self) -> HistoryCollectorData {
+        self.do_with_data(|data| std::mem::take(data))
```
I changed take_collected a bit and this no longer applies

cvybhu commented:
v2:
* Fixed a bunch of typos
* Removed `Default` implementation for `*Id` structs
* Used `take` in example
* Changed `take_collected` implementation - the old one set the IDs to 0, which could be bad if old events came in after taking it out. The new `take` implementation preserves the next `IDs` to assign.

cvybhu commented:
v4:
* Changed `match, continue` to if let
